src/main/java/de/tum/in/www1/artemis/service/connectors/localci/scaparser/strategy/PMDParser.java

```diff
@@ -1,71 +1,77 @@
 package de.tum.in.www1.artemis.service.connectors.localci.scaparser.strategy;
 
-import static de.tum.in.www1.artemis.service.connectors.localci.scaparser.utils.XmlUtils.getChildElements;
-
+import java.io.IOException;
 import java.util.ArrayList;
 import java.util.List;
 
-import org.w3c.dom.Document;
-import org.w3c.dom.Element;
+import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
+import com.fasterxml.jackson.dataformat.xml.XmlMapper;
+import com.fasterxml.jackson.dataformat.xml.annotation.JacksonXmlElementWrapper;
+import com.fasterxml.jackson.dataformat.xml.annotation.JacksonXmlProperty;
+import com.fasterxml.jackson.dataformat.xml.annotation.JacksonXmlText;
 
 import de.tum.in.www1.artemis.domain.enumeration.StaticCodeAnalysisTool;
+import de.tum.in.www1.artemis.service.dto.StaticCodeAnalysisIssue;
 import de.tum.in.www1.artemis.service.dto.StaticCodeAnalysisReportDTO;
-import de.tum.in.www1.artemis.service.dto.StaticCodeAnalysisReportDTO.StaticCodeAnalysisIssue;
 
-/**
- * Parser strategy for PMD reports.
- */
-class PMDParser implements ParserStrategy {
+@JsonIgnoreProperties(ignoreUnknown = true)
+record PMDReport(@JacksonXmlElementWrapper(useWrapping = false) @JacksonXmlProperty(localName = "file") List<FileViolation> files) {
+}
+
+@JsonIgnoreProperties(ignoreUnknown = true)
+record FileViolation(@JacksonXmlProperty(isAttribute = true, localName = "name") String fileName,
+
+        @JacksonXmlElementWrapper(useWrapping = false) @JacksonXmlProperty(localName = "violation") List<Violation> violations) {
```

Comment:
Do the localNames do not match the variable names on purpose? (Singular vs. Plural)

Reply:
Yes. We collect a list of violations (therefore the java name `violations`)  by collecting all sub-elements of type `violation`.

A PMD report file looks like this:
```
<file name="/opt/bambooagent/bamboo-agent-home/xml-data/build-dir/MDTESTCODE-ARTEMISTESTUSER1-JOB1/assignment/src/code/Client.java">
        <violation beginline="6" endline="6" begincolumn="1" endcolumn="26" rule="UnusedImports" ruleset="Best Practices" package="code" class="Client" externalInfoUrl="https://pmd.github.io/pmd-6.33.0/pmd_rules_java_bestpractices.html#unusedimports" priority="4">
            Avoid unused imports such as 'javax.swing.JFrame'
        </violation>
        <violation beginline="7" endline="7" begincolumn="1" endcolumn="19" rule="DontImportJavaLang" ruleset="Code Style" package="code" class="Client" externalInfoUrl="https://pmd.github.io/pmd-6.33.0/pmd_rules_java_codestyle.html#dontimportjavalang" priority="4">
            Avoid importing anything from the package java.lang
        </violation>
        <violation beginline="7" endline="7" begincolumn="1" endcolumn="19" rule="UnusedImports" ruleset="Best Practices" package="code" class="Client" externalInfoUrl="https://pmd.github.io/pmd-6.33.0/pmd_rules_java_bestpractices.html#unusedimports" priority="4">
            Avoid unused imports such as 'java.lang'
        </violation>
        <violation beginline="73" endline="73" begincolumn="18" endcolumn="27" rule="UnusedLocalVariable" ruleset="Best Practices" package="code" class="Client" method="createRandomDatesList" variable="randomDate" externalInfoUrl="https://pmd.github.io/pmd-6.33.0/pmd_rules_java_bestpractices.html#unusedlocalvariable" priority="3">
            Avoid unused local variables such as 'randomDate'.
        </violation>
</file>
````